### DISCUSSION TITLE
렉시컬 환경(Lexical Environment)에 대해 설명해 주세요

### DISCUSSION THREAD

**Discussion by airman5573** [woowacourse-study/2022-woowahan-bansanghwe#73] [JavaScript]
# Lexical Environment
## 간단한 정의
식별자와 식별자에 바인딩된 값, 그리고 상위 스코프에 대한 참조를 기록하는 자료구조로 실행 컨텍스트를 구성하는 컴포넌트다.  

## ECMA Spec에서의 정의 (feat. papago)
**A Lexical Environment is a specification type used to define the association of Identifiers to specific variables and functions based upon the lexical nesting structure of ECMAScript code.**   
=> 어휘 환경은 ECMA스크립트 코드의 어휘 중첩 구조(lexical nesting structure)를 기반으로 특정 변수와 함수에 대한 식별자의 연관성을 정의하는 데 사용되는 규격 유형이다.
=> lexical nesting structure ? 이게 뭐지?

**A Lexical Environment consists of an [Environment Record](https://262.ecma-international.org/6.0/#sec-environment-records) and a possibly null reference to an outer Lexical Environment.**   
=> 어휘 환경은 환경 레코드와 외부 어휘 환경에 대한 null 참조로 구성됩니다.
=> possibly null reference ?
=> outer Lexical Environment ?
=> 아까 위에서 nesting structure라고 했는데, outer라는것은 상위의 lexical environment를 가라키는 참조변수인가보다.
=> 그래서 possibly null reference라는건 상위의 lexical environment가 없을수도 있으니까, possibly null 이네!

**Usually a Lexical Environment is associated with some specific syntactic structure of ECMAScript code such as a FunctionDeclaration, a BlockStatement, or a Catch clause of a TryStatement and a new Lexical Environment is created each time such code is evaluated.**  
=> 일반적으로 Lexical Environment는 ECMA스크립트 코드의 특정 구문 구조(FunctionDeclaration, BlockStatement, TryStatement의 Catch 절 등)와 연관되어 있으며 이러한 코드가 평가될 때마다 새로운 Lexical Environment가 생성된다.
=> 코드가 평가될때마다 새로운 Lexical Environment가 생성된다고라..."평가"라는것은 "실행"을 의미하는건가?
=> 그러니까 `function a() { ... }` / `if () { ... }` / `try {  } catch () { ... }` 이런 코드를 만나면 Lexical Environment를 생성한다는 의미인것 같다.

**An [Environment Record](https://262.ecma-international.org/6.0/#sec-environment-records) records the identifier bindings that are created within the scope of its associated Lexical Environment. It is referred to as the Lexical Environment’s EnvironmentRecord**  
=> Lexical Environment안에 EnvironmentRecord가 있구나.
=> 그리고 이 EnvironmentRecord안에서 identifier와 variable을 맵핑(binding)시켜놓았다는 의미 같다.
=> 지금까지의 내용을 바탕으로 모양을 생각해보면

```javascript
LexicalEnvironment: {
    EnvironmentRecord: {
      // Identifier bindings
    },
    outer: <null>,
}
```
이런 모습이 나온다.

**The outer environment reference is used to model the logical nesting of Lexical Environment values. The outer reference of a (inner) Lexical Environment is a reference to the Lexical Environment that logically surrounds the inner Lexical Environment. An outer Lexical Environment may, of course, have its own outer Lexical Environment. A Lexical Environment may serve as the outer environment for multiple inner Lexical Environments. For example, if a FunctionDeclaration contains two nested FunctionDeclarations then the Lexical Environments of each of the nested functions will have as their outer Lexical Environment the Lexical Environment of the current evaluation of the surrounding function.**  
=> 외부 환경 참조는 사전 환경 값의 논리적 중첩을 모델링하는 데 사용됩니다. (내부) 어휘 환경의 외부 참조는 내부 어휘 환경을 논리적으로 둘러싼 어휘 환경을 참조하는 것이다. 외부 어휘 환경은 물론 자체적인 어휘 환경을 가질 수 있다. 어휘 환경은 여러 내부 어휘 환경의 외부 환경 역할을 할 수 있다. 예를 들어, 함수 선언에 두 개의 중첩된 함수 선언이 포함된 경우, 중첩된 각 함수의 어휘 환경은 외부 어휘 환경으로 주변 함수의 현재 평가에 대한 어휘 환경을 가집니다.
=> 코드로 보자면

```javascript
function outerFn() {
  ...
  function innerFn() {
  ...
  }
}

OuterLexicalEnvironment: {
  EnvironmentRecord: {
    // identifier bindings
  },
  outer: <null>
}

InnerLexicalEnvironment: {
  EnvironmentRecord: {
    // identifier bindings
  },
  outer: OuterLexicalEnvironment
}
```
이런 모습일것이다.

**A global environment is a Lexical Environment which does not have an outer environment. The global environment’s outer environment reference is null. A global environment’s EnvironmentRecord may be prepopulated with identifier bindings and includes an associated global object whose properties provide some of [the global environment](https://262.ecma-international.org/6.0/#sec-global-environment-records)’s identifier bindings. This global object is the value of a global environment’s this binding. As ECMAScript code is executed, additional properties may be added to the global object and the initial properties may be modified.**  
=> 글로벌 환경은 외부 환경이 없는 어휘 환경이다. 글로벌 환경의 외부 환경 참조가 null입니다. 글로벌 환경의 환경 레코드는 식별자 바인딩으로 미리 채워질 수 있으며 [글로벌 환경](https://262.ecma-international.org/6.0/#sec-global-environment-binding)의 식별자 바인딩 중 일부를 제공하는 속성을 가진 관련 글로벌 개체를 포함합니다. 이 전역 객체는 전역 환경의 이 바인딩의 값입니다. ECMA스크립트 코드가 실행되면 글로벌 객체에 추가 속성이 추가되고 초기 속성이 수정될 수 있습니다.  
=> 코드로 보면 
```javascript
GlobalLexicalEnvironment: {
  EnvironmentRecord: {
    window: WindowObject,
    ...
    // identifier bindings
  },
  outer: <null>
}
```
이런 모습일것이다.

**A module environment is a Lexical Environment that contains the bindings for the top level declarations of a Module. It also contains the bindings that are explicitly imported by the Module. The outer environment of a module environment is a global environment.**  
=> 모듈 환경은 모듈의 최상위 선언에 대한 바인딩을 포함하는 사전 환경입니다. 또한 모듈에서 명시적으로 가져온 바인딩도 포함되어 있습니다. 모듈 환경의 외부 환경은 글로벌 환경이다.    
=> 코드로 보자면  
```javascript
import { sayHello } from 'speak.js';

...

ModuleLexicalEnvironment: {
  EnvironmentRecord: {
    sayHello: SayHelloFunction
    ...
    // identifier bindings
  },
  outer: GlobalLexicalEnvironment
}
```
이런 모양일것이다.

**A function environment is a Lexical Environment that corresponds to the invocation of an [ECMAScript function object](https://262.ecma-international.org/6.0/#sec-ecmascript-function-objects). A function environment may establish a new this binding. A function environment also captures the state necessary to support super method invocations.**  
=> 함수 환경은 [ECMAScript function object](https://262.ecma-international.org/6.0/#sec-ecmascript-function-function-function-function) 호출에 해당하는 어휘 환경이다. 함수 환경은 this 바인딩을 새로 설정할 수 있다. 함수 환경은 또한 슈퍼 메소드 호출을 지원하는 데 필요한 상태를 캡처한다.  
=> invocation of an ECMAScript function object 이라는것은 함수를 호출한다는 의미다. 그래서 함수를 호출할때 function environment가 생성된다는 의미다.

**Lexical Environments and [Environment Record](https://262.ecma-international.org/6.0/#sec-environment-records) values are purely specification mechanisms and need not correspond to any specific artefact of an ECMAScript implementation. It is impossible for an ECMAScript program to directly access or manipulate such values.**  
=> 어휘 환경 및 [환경 기록](https://262.ecma-international.org/6.0/#sec-환경 기록) 값은 순전히 사양 메커니즘이며 ECMA스크립트 구현의 특정 아티팩트에 해당할 필요는 없습니다. ECMA스크립트 프로그램이 이러한 값에 직접 접근하거나 조작하는 것은 불가능하다.  
=> 그러니까 그냥 이런 식으로 작동한다는 spec일 뿐이고, 구현은 엔진 마음대로 구현해도 된다는 의미이다.  

# 참고
- https://blog.bitsrc.io/understanding-execution-context-and-execution-stack-in-javascript-1c9ea8642dd0
- https://segmentfault.com/a/1190000040045584/en

**Comment by airman5573** [comment]
스코프와 Lexical Environment는 어떤 관계인가요?

**Comment by airman5573** [comment]
Lexical Environment와 클로저는 어떤 관계인가요?

**Comment by byhhh2** [comment]
environment record란?

**Comment by byhhh2** [comment]
lexical environment와 environment record가 왜 필요하고 어디에 이용되나요?

**Comment by byhhh2** [comment]
스코프는 어떻게 동작하나요 (with lexical environment)

**Comment by byhhh2** [comment]
스코프 체인과 lexical environment가 어떻게 동작하는지
